main.cpp

```diff
@@ -274,395 +223,238 @@ struct HeapA
+//PART 7 CODE START -------------------------------------------------------------
 
-private:
-    float* value;
-    FloatType& powInternal( float val );
-};
+template<typename NumericType>       
 
-//======================================================
-struct DoubleType
+struct Numeric
 {
+    using Type = NumericType;     
 
-    DoubleType(double hDVal) : value(new double (hDVal) ){}
-    ~DoubleType()
-    {
-        delete value;
-        value = nullptr;
-    }
-
-    DoubleType& operator+=(double rhs);
-    DoubleType& operator-=(double rhs);
-    DoubleType& operator*=(double rhs);
-    DoubleType& operator/=(double rhs);
-
-    operator double() const {return *value;}
-
-    DoubleType& pow( double val );
-    DoubleType& pow( const FloatType& ft );
-    DoubleType& pow( const DoubleType& dt );
-    DoubleType& pow( const IntType& it );
-
-    DoubleType& apply( std::function<DoubleType&(double&)> func );
-    DoubleType& apply( void(*funcPtr)(double&) );
+    Numeric(Type val) : value( std::make_unique<Type>(val) ) { }
 
-private:
-    double* value;
-    DoubleType& powInternal( double value);
-
-};
-//======================================================
-struct IntType
-{
+    operator Type() const { return *value; }
 
-    IntType(int hIVal) : value(new int (hIVal) ){}
-    ~IntType()
+    Numeric& operator+=( const Type rhs )
     {
-        delete value;
-        value = nullptr;
+        *value += rhs;
+        return *this;
     }
 
-    IntType& operator+=( int rhs );
-    IntType& operator-=( int rhs );
-    IntType& operator*=( int rhs );
-    IntType& operator/=( int rhs );
-
-    operator int() const { return *value; }
-
-    IntType& pow( int val );
-    IntType& pow( const FloatType& ft );
-    IntType& pow( const DoubleType& dt );
-    IntType& pow( const IntType& it );
-
-    IntType& apply( std::function<IntType&(int&)> func );
-    IntType& apply( void(*funcPtr)(int&) );
-
-private:
-    int* value;
-    IntType& powInternal( int val );
-};
-
-struct Point
-{
-    Point(const IntType& x_, const IntType& y_) : x(static_cast<float>(x_)), y(static_cast<float>(y_)){}    
-    Point(const FloatType& x_, const FloatType& y_) : x(static_cast<float>(x_)), y(static_cast<float>(y_)){}
-    Point(const DoubleType& x_, const DoubleType& y_) : x(static_cast<float>(x_)), y(static_cast<float>(y_)){}
-
-    Point& multiply(float m)
+    Numeric& operator-=( const Type rhs )
     {
-        x *= m;
-        y *= m;
+        *value -= rhs;
         return *this;
     }
 
-    Point& multiply( const FloatType& m )
+    Numeric& operator*=( const Type rhs )
     {
-        return multiply(static_cast<float>(m));
+        *value *= rhs;
+        return *this;
     }
-    
-    Point& multiply( const DoubleType& m )
+
+    template<typename DType>
+    Numeric& operator/=(const DType& rhs)
     {
-        return multiply(static_cast<float>(m));
+        if constexpr (std::is_same<Type, int>::value)
```

Comment:
FYI, Your class template's type is not `Type`.  This will come back to bite you on the next project part.